Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp

```diff
@@ -1960,15 +1963,87 @@ auto LLIntGenerator::addI31GetU(ExpressionType ref, ExpressionType& result) -> P
+{
+    if (args.isEmpty()) {
+        result = VirtualRegister();
+        return;
+    }
+
+    // Shift all the arguments up the stack by one slot (for example, because the first
+    // argument slot was just designated for the return value of the struct_new operation.
+    // Example: args = { loc18, loc19, loc20 }
+    // We assume the args have been popped off the stack. So result = loc18
+
+    // So we start by allocating a vector of stack slots for the struct fields.
+    // In the example, destinations = { loc19, loc20, loc21 }
+    // We do this before moving the arguments because otherwise, we would
+    // overwrite one of the existing slots before it's copied.
+    // That means in the example, we would overwrite loc19 before it's used.
+    Vector<ExpressionType> destinations(args.size());
+    for (unsigned i = 0; i < args.size(); i++)
+        destinations[i] = push(NoConsistencyCheck); // We skip the consistency check because at this point, the result slot overlaps with the first argument
+
+    // By similar reasoning to the `addThrow` operation:
+    // We have to materialize the arguments here since it might include constants or
+    // delayed moves, but the wasm_struct_new opcode expects all the arguments to be contiguous
+    // in the stack.
+
+    // Iterate in reverse order so that we never overwrite a register before it's used
+    // In the example, we would start with: loc21 := loc20, which is valid since
+    // loc21 is uninitialized
+    for (int i = args.size() - 1; i >= 0; --i) {
+        auto& arg = args[i];
+        ExpressionType argLoc = destinations[i];
+        WasmMov::emit(this, argLoc, arg);
+        arg = argLoc;
+    }
+
+    // Now the last arg will be at the highest virtual register index, which will have the lowest address.
+    // We set the struct_new instruction's `m_firstValue` parameter to the last arg, but the arguments are in reverse order.
+    // In the example, `m_firstValue` would be loc21, which contains the argument that was on top of the stack (the last one).
+    // The implementation of the operation that consumes these arguments has to iterate over its arguments in reverse order.
+    // (In the example, arguments = { loc21, loc20, loc19 })
+    result = args.last();
+}
+
+void LLIntGenerator::adjustAndCheckStack(unsigned args)
+{
+    // Adjust stack to reflect popping arguments
+    m_stackSize -= args;
+
+    // This assumes that exactly one result has been pushed onto the control stack
+    // but not the expression stack yet; so we temporarily pretend it's
+    // not on the control stack by decrementing the stack size
+    m_stackSize--;
+    checkConsistency();
+    m_stackSize++;
+}
+
+auto LLIntGenerator::addArrayNewFixed(uint32_t index, Vector<ExpressionType>& args, ExpressionType& result) -> PartialResult
+{
+    result = push();
+
+    ExpressionType firstValue;
+    shiftOperands(args, firstValue);
```

Comment:
Hmm, since we end up calling a C thunk in the end anyway, this seems kinda bad. I guess I can't really think of a better way though right now.

Reply:
This looks very dangerous to me. Why not making it LLInt builtin calls?

Reply:
This follows the structure of `addStructNew()` (or at least it will once https://github.com/WebKit/WebKit/pull/10342 lands), which also calls a C thunk in the end. I was assuming that if there was a better way, it would have been used for struct.new as well (since both take a variable number of arguments), but I'm open to suggestions.

Reply:
@Constellation I don't see any examples of existing builtins that either take a variable number of arguments or an array of arguments -- how would the arguments be passed if not on the stack?

Reply:
I guess the real solution here is that we would normally write this sort of thing directly in the llint, and it would get() them directly. 

As for today, I guess this is fine since it doesn't actually "shift" so much as copy, but it is certainly suboptimal. It shares the same problem as the existing code though, yes.

Reply:
Waiting for your feedback on #10342 to see whether I should use the same approach here as I just did there.